server/SessionService.java

```diff
@@ -95,28 +99,39 @@ public synchronized void resetIdleTimeout() {
     }
 
     private void cancelIdleTimeout() {
-        assert idleTimeoutTask != null;
         idleTimeoutTask.cancel(false);
     }
 
     private synchronized void mayStartIdleTimeout() {
-        if (isOpen() && transactionServices.isEmpty()) {
-            idleTimeoutTask = scheduled().schedule(this::idleTimeout, options.sessionIdleTimeoutMillis(), MILLISECONDS);
+        if (isOpen() && transactionServices.isEmpty() && !idleTimeoutActive()) {
+            startIdleTimeout();
         }
     }
 
+    private boolean idleTimeoutActive() {
+        return !idleTimeoutTask.isDone();
+    }
+
     private void idleTimeout() {
-        if (transactionServices.isEmpty()) {
-            close();
-            LOG.warn("Session with ID " + session.uuid() + " timed out due to inactivity");
-        } else resetIdleTimeout();
+        try {
+            accessLock.writeLock().lock();
+            if (transactionServices.isEmpty()) {
+                if (isOpen.compareAndSet(true, false)) {
+                    session.close();
+                    typeDBSvc.closed(this);
+                    LOG.warn("Session with ID " + session.uuid() + " timed out due to inactivity");
+                }
+            }
```

Comment:
what about if the `transactionServices` are not empty?

Reply:
Then the transaction must have been opened between entering the method and acquiring the lock. We do nothing and release the lock.

Reply:
Yes, it would try to cancel a timeout, which wouldn't do anything, and try to restart it. The last transaction may have been closed by the time we reach this line after checking the condition, so we also might cancel a new active timeout and replace it with a new future. Either way, essentially a very busy no-op.

Reply:
right - i actually meant we could delete the method then!

Reply:
No, it's used in TypeDBService (L215)

Reply:
ah okay